tests/unit/test_decide_user_install.py

```diff
@@ -0,0 +1,124 @@
+                  filter(lambda args: sum(map(bool, args)) > 1,
+                         product((False, True), (None, 'foo'),
+                                 (None, 'bar'), (None, 'baz'))))
+def test_conflicts(use_user_site, prefix_path, target_dir, root_path):
+    """Test conflicts of target, user, root and prefix options."""
+    with raises(CommandError):
+        decide_user_install(
+            use_user_site=use_user_site, prefix_path=prefix_path,
+            target_dir=target_dir, root_path=root_path)
+
+
+def test_target_exists(monkeypatch):
+    """Test exist target path."""
+    monkeypatch.setattr(WRITABLE, true)
+    monkeypatch.setattr(EXISTS, true)
+    monkeypatch.setattr(ISDIR, false)
+    with raises(InstallationError):
+        decide_user_install(target_dir='bar')
+    for exists, isdir in (false, false), (false, true), (true, true):
+        monkeypatch.setattr(EXISTS, exists)
+        monkeypatch.setattr(ISDIR, isdir)
+        assert decide_user_install(target_dir='bar') is False
+
+
+def test_target_writable(monkeypatch):
+    """Test writable path check with target specified."""
+    monkeypatch.setattr(EXISTS, false)
+    monkeypatch.setattr(WRITABLE, false)
+    with raises(InstallationError):
+        decide_user_install(target_dir='bar')
+    monkeypatch.setattr(WRITABLE, true)
+    assert decide_user_install(target_dir='bar') is False
+
+
+def test_prefix_writable(monkeypatch):
+    """Test writable path check with prefix specified."""
+    monkeypatch.setattr(WRITABLE, false)
+    with raises(InstallationError):
+        decide_user_install(prefix_path='foo')
+    monkeypatch.setattr(WRITABLE, true)
+    assert decide_user_install(prefix_path='foo') is False
+
+
+def test_global_site_writable(monkeypatch):
+    """Test error handling and user site-packages decision
+    with writable and non-writable global site-packages.
+    """
+    monkeypatch.setattr(VIRTUALENV_NO_GLOBAL, false)
+    monkeypatch.setattr(ENABLE_USER_SITE, True)
+    monkeypatch.setattr(WRITABLE, false)
+    with raises(InstallationError):
+        decide_user_install(use_user_site=False)
+    with raises(InstallationError):
+        decide_user_install(root_path='baz')
+    assert decide_user_install() is True
+
+    monkeypatch.setattr(WRITABLE, true)
+    assert decide_user_install(use_user_site=True) is True
+    assert decide_user_install(root_path='baz') is False
+    assert decide_user_install() is False
+
+
+def test_enable_user_site(monkeypatch):
+    """Test with different values of site.ENABLE_USER_SITE."""
+    monkeypatch.setattr(VIRTUALENV_NO_GLOBAL, false)
+    monkeypatch.setattr(ENABLE_USER_SITE, None)
+    monkeypatch.setattr(WRITABLE, true)
+    assert decide_user_install() is False
+    assert decide_user_install(use_user_site=False) is False
+    with raises(InstallationError):
+        decide_user_install(use_user_site=True)
+
+    monkeypatch.setattr(ENABLE_USER_SITE, False)
+    assert decide_user_install() is False
```

Comment:
It looks like you mix a lot of distinct tests in one test function, I think they should be separate, with separate names/descriptions. You can move the shared "setup" code to fixtures so that it's not repeated and also separated from testing.

Reply:
I'm not sure if I get what you mean: is it to separate test cases for `site.ENABLE_USER_SITE` being {None, False, True}?  BTW I've just cleaned up the setups and hope it looks clearer what's being test now.

Reply:
Ideally, each test function should contain one `assert` or `with raises`. In your case, it's collections of assertions that test something. So yes, I'd separate by `ENABLE_USER_SITE`.

There's a few reasons for this:
1) The first assert that fails will interrupt this test function.
2) You need other tests to still be performed.
3) In the test run report, these tests should show up separately with separate statuses.

P.S. If you're able to come up with a good `parametrize` definition, that'll also work for me: parametrized tests show up separately.

Reply:
Thanks for the insight!